File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoAudit/FanoutSink.swift
import Foundation

/// Forwards events to multiple sinks. Errors in one sink do not poison others.
public final class FanoutSink: Sink {
    private let sinks: [Sink]

    public init(_ sinks: [Sink]) {
        self.sinks = sinks
    }

    public convenience init(_ sinks: Sink...) {
        self.init(sinks)
    }

    public func emit(_ event: Event) {
        for s in sinks {
            s.emit(event)
        }
    }
}

File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/audit/fanout_sink.rb
# frozen_string_literal: true

module IAIso
  module Audit
    # Broadcasts events to multiple sinks. Errors in any one sink do
    # not poison the others.
    class FanoutSink
      def initialize(*sinks)
        @sinks = sinks.flatten.freeze
      end

      def emit(event)
        @sinks.each do |s|
          begin
            s.emit(event)
          rescue StandardError
            # swallow
          end
        end
      end
    end
  end
end
